lib/mdx.ml

```diff
@@ -73,31 +78,103 @@ let parse l =
+             ; labels = []
+             ; header = Some "ocaml"
+             ; contents = Compat.String.split_on_char '\n' code_block.contents
+             ; value = Raw
+             }
+         in
+         cursor := code_block.location.end_;
+         [ pre_text; Text "{["; block; Text "]}" ])
+      code_blocks
+    |> List.concat
+  in
+  let eof =
+    let lines = Compat.String.split_on_char '\n' file_contents in
+    { Odoc_model.Location_.line = List.length lines
+    ; column = String.length (List.rev lines |> List.hd)
+    }
+  in
+  let eof_is_beyond_location (loc : Odoc_model.Location_.point) =
+    eof.line > loc.line || (eof.line = loc.line && eof.column > loc.column)
+  in
+  if eof_is_beyond_location !cursor
+  then (
+    let remainder = Mli_parser.slice file_contents ~start:!cursor ~end_:eof in
+    if not (Compat.String.equal remainder "") then tokens @ [ Text remainder ] else tokens)
+  else tokens
+
 type syntax = Syntax.t =
   | Normal
   | Cram
+  | Mli
+
+let parse_lexbuf file_contents syntax l =
+  match syntax with
+  | Syntax.Mli -> parse_mli file_contents
+  | Syntax.Normal | Syntax.Cram -> parse (Lexer_mdx.token syntax l)
 
-let parse_lexbuf syntax l = parse (Lexer.token syntax l)
-let parse_file syntax f = parse (Lexer.token syntax (snd (Misc.init f)))
-let of_string syntax s = parse_lexbuf syntax (Lexing.from_string s)
+let parse_file syntax f = parse (Lexer_mdx.token syntax (snd (Misc.init f)))
 
-let eval = function
+let of_string syntax s =
+  match syntax with
+  | Syntax.Mli -> parse_mli s
+  | Syntax.Normal | Syntax.Cram -> parse_lexbuf s syntax (Lexing.from_string s)
+
+let eval syntax = function
   | Section _ | Text _ as x -> x
   | Block t as x ->
-    let t' = Block.eval t in
+    let t' = Block.eval syntax t in
     if t == t' then x else Block t'
 
 type expect_result =
   | Identical
   | Differs
 
+let remove_whitespace str =
+  let open Astring in
+  List.fold_right (fun s acc ->
+    String.cuts ~sep:s acc |> String.concat ~sep:""
+  ) [ " "; "\n"; "\t" ] str
+
 let run_str ~syntax ~f file =
   let file_contents, lexbuf = Misc.init file in
-  let items = parse_lexbuf syntax lexbuf in
-  let items = List.map eval items in
+  let items = parse_lexbuf file_contents syntax lexbuf in
+  let items = List.map (fun i -> eval syntax i) items in
   Log.debug (fun l -> l "run @[%a@]" dump items);
   let corrected = f file_contents items in
-  let result = if corrected <> file_contents then Differs else Identical in
+  (* The code blocks in mli syntax are often formatted by a tool like ocp-indent or
+     ocamlformat, and it would be impossible to match this formatting exactly down to
+     every space and newline.
+
+     Instead, for this syntax only, we make the diff whitespace-invariant.
+  *)
+  let has_changed = match syntax with
+    | Syntax.Cram | Syntax.Normal -> corrected <> file_contents
+    | Syntax.Mli -> remove_whitespace corrected <> remove_whitespace file_contents
```

Comment:
I'm a bit worried about removing all whitespaces for the diff here. It should work now because we only support toplevel blocks but I feel like it's easy to forget about this and break mli support if we had support for other blocks in mli files.

We don't necessarily have to fix this now but if we merge it as is we must open an issue as a reminder that we need to imrove that.

In particular, what I have in mind is tied to a refactoring of the tests logic where we would better split I/O from the generation of the updated blocks. Once we have that we can easily diff individual blocks and keep track of any a change there instead of diffing the whole document as a string. cc @gpetiot @voodoos.

Reply:
This is important enough to our use case here that I think I will aim to leave it in -- and then address later along the lines you suggest.

Reply:
Sounds good, I'll add an issue once we merge!